projects/PM_hash_navigation/snapshot_db.h

```diff
@@ -146,13 +148,11 @@ class Snapshot_DB
         std::vector<cv::Mat> rotatedViews;
         // get all rotations of the current view
         for (int i = 0; i < currentView.size().width; i++) {
-            cv::Mat out = currentView.clone();
-            cv::Mat currCopy = currentView.clone();
-            rollImage(currCopy, out, i);
+            rotatedViews.emplace_back(currentView.size(), currentView.type());
+            rollImage(currentView, rotatedViews.back(), i);
             unsigned long long int rotHash;
-            m_hasher.computeDCT_Hash(out, rotHash);
+            m_hasher.computeDCT_Hash(rotatedViews.back(), rotHash);
             rotationHashes.push_back(rotHash);
-            rotatedViews.push_back(out);
```

Comment:
I've removed some more in-loop allocations here. You don't want to repeatedly make copies of ``currentView`` if you can possibly avoid it.

Reply:
In an ideal world you don't really want to regenerate these ``std::vector``s every time the function is called either. ``rotatedViews`` could be populated once in the constructor and its contents could be reused and the other ``std::vector``s could also be members and then ``clear()``ed every time the function is called.

Reply:
I have clearly suceeded in my quest to make @alexdewar deeply afraid of heap allocations 🥳

Reply:
:laughing: And not just that! I now have nightmares about const correctness :fearful:

Reply:
and thus the C++ cargo cult continues :smile: